### PR TITLE
Sgd support update selected rows

### PR DESCRIPTION
**Pull request by jacquesqiao**
task list: https://github.com/PaddlePaddle/Paddle/issues/9211

### CODE REVIEW

File: paddle/fluid/operators/sgd_op.cc [hunk 1]
```diff
@@ -43,20 +43,32 @@ class SGDOp : public framework::OperatorWithKernel {
...
+          framework::ToDataType(table_var->Get<framework::LoDTensor>().type()),
+          ctx.device_context());
+    } else if (table_var->IsType<framework::SelectedRows>()) {
+      return framework::OpKernelType(
+          framework::ToDataType(
+              table_var->Get<framework::SelectedRows>().value().type()),
+          ctx.device_context());
+    } else {
+      PADDLE_THROW("Param should be LoDTensor or SelectedRows");
+    }
```
Comment (chengduoZH) [inline paddle/fluid/operators/sgd_op.cc]: line46-58 is similar to line21-35 of lookup_tabel_op, so I think we should add a new API, like `GetPODTypeFromVar`, do you think so?

File: paddle/fluid/operators/sgd_op.h [hunk 2]
```diff
@@ -23,60 +23,97 @@ namespace operators {
...
+        }
+      } else {
+        PADDLE_THROW("Unsupported Variable Type of Grad");
+      }
+    } else if (param_var->IsType<framework::SelectedRows>()) {
+      PADDLE_ENFORCE(grad_var->IsType<framework::SelectedRows>(),
+                     "when param "
+                     "is SelectedRows, gradient should also be SelectedRows");
+      const auto &param = param_var->Get<framework::SelectedRows>();
+      auto *param_out = ctx.Output<framework::SelectedRows>("ParamOut");
```
Comment (chengduoZH) [inline paddle/fluid/operators/sgd_op.h]: Can `param_out` and `param` be different?

Comment (jacquesqiao) [reply paddle/fluid/operators/sgd_op.h]: Currently, they need to share the same memory, but we should use different vars to represent them for:
1. Input is const, the output is mutable.
2. In the future, they can be different.

Comment (chengduoZH) [reply paddle/fluid/operators/sgd_op.h]: Now that, it should have a check under line89, like: `PADDLE_ENFORCE_EQ(param, param_out);`

File: paddle/fluid/operators/sgd_op.cc (hunk 1 above)
Comment (jacquesqiao) [reply paddle/fluid/operators/sgd_op.cc]: done

File: paddle/fluid/operators/sgd_op.h (hunk 2 above)
Comment (jacquesqiao) [reply paddle/fluid/operators/sgd_op.h]: I think this enforce is not needed.

**Pull request merged**